File: crates/tbdex/src/resources/offering.rs
use super::{ResourceKind, ResourceMetadata, Result};
use crate::{
    json::{FromJson, ToJson},
    json_schemas::generated::{OFFERING_DATA_JSON_SCHEMA, RESOURCE_JSON_SCHEMA},
    DEFAULT_PROTOCOL_VERSION,
};
use chrono::Utc;
use serde::{Deserialize, Serialize};
use web5::{
    credentials::presentation_definition::PresentationDefinition, dids::bearer_did::BearerDid,
};

/// Represents an Offering resource in the tbDEX protocol.
///
/// An Offering resource is created by a PFI to define the requirements for a given exchange,
/// including payin/payout details, rates, and other conditions.
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
pub struct Offering {
    /// Metadata about the resource, including sender, type of resource, and protocol information.
    pub metadata: ResourceMetadata,

    /// The public data part of the Offering, such as exchange rates and payment methods.
    pub data: OfferingData,

    /// The signature verifying the authenticity and integrity of the Offering resource.
    pub signature: String,
}

impl ToJson for Offering {}
impl FromJson for Offering {}

impl Offering {
    /// Creates a new Offering resource.
    ///
    /// # Arguments
    ///
    /// * `from` - The DID of the sender (the PFI).
    /// * `data` - The data containing the offering details.
    /// * `protocol` - Optional protocol version; defaults to the current version if not provided.
    ///
    /// # Returns
    ///
    /// A new instance of `Offering` containing the metadata, data, and an empty signature.
    pub fn create(from: &str, data: &OfferingData, protocol: Option<String>) -> Result<Self> {
        let now = Utc::now().to_rfc3339();

        let metadata = ResourceMetadata {
            kind: ResourceKind::Offering,
            from: from.to_string(),
            id: ResourceKind::Offering.typesafe_id()?,
            protocol: protocol.unwrap_or_else(|| DEFAULT_PROTOCOL_VERSION.to_string()),
            created_at: now.clone(),
            updated_at: Some(now),
        };

        let offering = Self {
            metadata: metadata.clone(),
            data: data.clone(),
            signature: String::default(),
        };

        Ok(offering)
    }

    /// Signs the Offering resource using the provided Bearer DID.
    ///
    /// # Arguments
    ///
    /// * `bearer_did` - The DID to sign the Offering resource.
    ///
    /// # Returns
    ///
    /// An empty result, or an error if the signing process fails.
    pub fn sign(&mut self, bearer_did: &BearerDid) -> Result<()> {
        self.signature = crate::signature::sign(
            bearer_did,
            &serde_json::to_value(&self.metadata)?,
            &serde_json::to_value(&self.data)?,
        )?;
        Ok(())
    }

    /// Verifies the validity of the Offering resource.
    ///
    /// This method ensures that the resource adheres to its JSON schema
    /// and verifies the signature to ensure authenticity and integrity.
    ///
    /// # Returns
    ///
    /// An empty result if verification succeeds, or an error if verification fails.
    pub fn verify(&self) -> Result<()> {
        // verify resource json schema
        crate::json_schemas::validate_from_str(RESOURCE_JSON_SCHEMA, self)?;

        // verify data json schema
        crate::json_schemas::validate_from_str(OFFERING_DATA_JSON_SCHEMA, &self.data)?;

        // verify signature
        crate::signature::verify(
            &self.metadata.from,
            &serde_json::to_value(self.metadata.clone())?,
            &serde_json::to_value(self.data.clone())?,
            &self.signature,
        )?;

        Ok(())
    }
}

/// Represents the data for an Offering resource in the tbDEX protocol.
///
/// This includes details such as the exchange rate, payin/payout methods, and any required credentials.
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct OfferingData {
    /// A brief description of what is being offered.
    pub description: String,

    /// The exchange rate, indicating how many payout units are provided per payin unit.
    pub payout_units_per_payin_unit: String,

    /// Details about the payin currency and available methods.
    pub payin: PayinDetails,

    /// Details about the payout currency and available methods.
    pub payout: PayoutDetails,

    /// Optional required claims (Verifiable Credentials) needed to participate in the offering.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub required_claims: Option<PresentationDefinition>,

    /// Details about the offering's cancellation policy.
    pub cancellation: CancellationDetails,
}

/// Represents the details of the payin for an Offering.
///
/// This includes the currency code, optional limits, and available methods for making the payin.
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Default)]
#[serde(rename_all = "camelCase")]
pub struct PayinDetails {
    /// The currency code (ISO 4217 format) for the payin.
    pub currency_code: String,

    /// Optional minimum payin amount.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min: Option<String>,

    /// Optional maximum payin amount.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max: Option<String>,

    /// A list of available methods for making the payin.
    pub methods: Vec<PayinMethod>,
}

/// Represents a method for making a payin in an Offering.
///
/// This includes details about the method, such as its kind, optional name and description, and any associated fees.
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Default)]
#[serde(rename_all = "camelCase")]
pub struct PayinMethod {
    /// The unique identifier for the payment method.
    pub kind: String,

    /// Optional name of the payment method.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,

    /// Optional description of the payment method.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,

    /// Optional group categorization for the payment method.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub group: Option<String>,

    /// Optional JSON schema specifying required payment details.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub required_payment_details: Option<serde_json::Value>,

    /// Optional fee associated with the payment method.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fee: Option<String>,

    /// Optional minimum amount for using the payment method.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min: Option<String>,

    /// Optional maximum amount for using the payment method.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max: Option<String>,
}

/// Represents the details of the payout for an Offering.
///
/// This includes the currency code, optional limits, and available methods for receiving the payout.
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Default)]
#[serde(rename_all = "camelCase")]
pub struct PayoutDetails {
    /// The currency code (ISO 4217 format) for the payout.
    pub currency_code: String,

    /// Optional minimum payout amount.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min: Option<String>,

    /// Optional maximum payout amount.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max: Option<String>,

    /// A list of available methods for receiving the payout.
    pub methods: Vec<PayoutMethod>,
}

/// Represents a method for receiving a payout in an Offering.
///
/// This includes details about the method, such as its kind, optional name and description, and any associated fees.
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct PayoutMethod {
    /// The unique identifier for the payout method.
    pub kind: String,

    /// Optional name of the payout method.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,

    /// Optional description of the payout method.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,

    /// Optional group categorization for the payout method.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub group: Option<String>,

    /// Optional JSON schema specifying required payment details.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub required_payment_details: Option<serde_json::Value>,

    /// Optional fee associated with the payout method.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fee: Option<String>,

    /// Optional minimum amount for using the payout method.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min: Option<String>,

    /// Optional maximum amount for using the payout method.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max: Option<String>,

    /// The estimated time (in seconds) for the payout to be settled.
    pub estimated_settlement_time: i64,
}

/// Represents the cancellation policy for an Offering.
///
/// This includes whether cancellation is enabled and optional terms describing the cancellation policy.
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Default)]
#[serde(rename_all = "camelCase")]
pub struct CancellationDetails {
    /// Whether cancellation is enabled for the offering.
    pub enabled: bool,

    /// Optional URL to a page that describes the terms of cancellation.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub terms_url: Option<String>,

    /// Optional human-readable description of the cancellation terms.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub terms: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use web5::dids::methods::did_jwk::DidJwk;

    #[test]
    fn can_create_and_sign_and_verify() {
        let bearer_did = DidJwk::create(None).unwrap();

        let mut offering = Offering::create(
            &bearer_did.did.uri,
            &OfferingData {
                description: "Selling BTC for USD".to_string(),
                payout_units_per_payin_unit: "1.5".to_string(),
                payin: PayinDetails {
                    currency_code: "USD".to_string(),
                    ..Default::default()
                },
                payout: PayoutDetails {
                    currency_code: "BTC".to_string(),
                    ..Default::default()
                },
                required_claims: Some(PresentationDefinition {
                    id: "7ce4004c-3c38-4853-968b-e411bafcd945".to_string(),
                    name: None,
                    purpose: None,
                    input_descriptors: vec![],
                }),
                cancellation: CancellationDetails {
                    enabled: false,
                    ..Default::default()
                },
            },
            None,
        )
        .unwrap();

        offering.sign(&bearer_did).unwrap();

        assert_ne!(String::default(), offering.signature);

        let offering_json_string = offering.to_json_string().unwrap();

        assert_ne!(String::default(), offering_json_string);

        let parsed_offering = Offering::from_json_string(&offering_json_string).unwrap();

        assert_eq!(offering, parsed_offering);
    }
}

#[cfg(test)]
mod tbdex_test_vectors_protocol {
    use super::*;
    use std::fs;

    #[derive(Debug, serde::Deserialize)]
    pub struct TestVector {
        pub input: String,
        pub output: Offering,
    }

    #[test]
    fn parse_offering() {
        let path = "../../tbdex/hosted/test-vectors/protocol/vectors/parse-offering.json";
        let test_vector_json: String = fs::read_to_string(path).unwrap();

        let test_vector: TestVector = serde_json::from_str(&test_vector_json).unwrap();
        let parsed_offering: Offering = Offering::from_json_string(&test_vector.input).unwrap();

        assert_eq!(test_vector.output, parsed_offering);
    }
}

File: bound/kt/src/main/kotlin/tbdex/sdk/resources/Offering.kt
package tbdex.sdk.resources

import tbdex.sdk.Json
import tbdex.sdk.TbdexException
import tbdex.sdk.rust.Offering as RustCoreOffering
import tbdex.sdk.rust.fromWeb5
import tbdex.sdk.rust.BearerDid as RustCoreBearerDid
import web5.sdk.dids.BearerDid
import web5.sdk.vc.pex.PresentationDefinition

/**
 * Represents an Offering resource in the tbDEX protocol.
 *
 * An Offering is created by a PFI to define the exchange parameters, such as payin/payout details,
 * exchange rates, required credentials, and cancellation policies.
 *
 * @property metadata Metadata about the resource, including sender and resource type.
 * @property data The details of the offering, including payin/payout methods, required claims, and cancellation.
 * @property signature The signature verifying the authenticity and integrity of the Offering resource.
 * @property rustCoreOffering The underlying RustCore representation of the Offering resource.
 */
data class Offering private constructor(
    val metadata: ResourceMetadata,
    val data: OfferingData,
    val signature: String,
    internal val rustCoreOffering: RustCoreOffering
) {
    companion object {
        /**
         * Creates a new Offering resource.
         *
         * @param from The DID of the sender (the PFI).
         * @param data The offering details including exchange rates, payin/payout details, and required claims.
         * @param protocol Optional protocol version.
         * @return The newly created Offering resource.
         * @throws TbdexException if the creation process fails.
         */
        fun create(
            from: String,
            data: OfferingData,
            protocol: String? = null
        ): Offering {
            try {
                val jsonSerializedData = Json.stringify(data)
                val rustCoreOffering = RustCoreOffering.create(from, jsonSerializedData, protocol)
                val rustCoreData = rustCoreOffering.getData()
                return Offering(
                    ResourceMetadata.fromRustCore(rustCoreData.metadata),
                    Json.jsonMapper.readValue(rustCoreData.jsonSerializedData, OfferingData::class.java),
                    rustCoreData.signature,
                    rustCoreOffering
                )
            } catch (e: tbdex.sdk.rust.TbdexException.Exception) {
                throw TbdexException.fromRustCore(e)
            }
        }

        /**
         * Parses an Offering resource from a JSON string.
         *
         * @param json The JSON string representing the Offering resource.
         * @return The deserialized Offering resource.
         * @throws TbdexException if parsing fails.
         */
        fun fromJsonString(json: String): Offering {
            try {
                val rustCoreOffering = RustCoreOffering.fromJsonString(json)
                val rustCoreData = rustCoreOffering.getData()
                val data = Json.jsonMapper.readValue(rustCoreOffering.getData().jsonSerializedData, OfferingData::class.java)
                return Offering(
                    ResourceMetadata.fromRustCore(rustCoreData.metadata),
                    data,
                    rustCoreData.signature,
                    rustCoreOffering
                )
            } catch (e: tbdex.sdk.rust.TbdexException.Exception) {
                throw TbdexException.fromRustCore(e)
            }
        }

        internal fun fromRustCoreOffering(rustCoreOffering: RustCoreOffering): Offering {
            val rustCoreData = rustCoreOffering.getData()
            val data = Json.jsonMapper.readValue(rustCoreOffering.getData().jsonSerializedData, OfferingData::class.java)
            return Offering(
                ResourceMetadata.fromRustCore(rustCoreData.metadata),
                data,
                rustCoreData.signature,
                rustCoreOffering
            )
        }
    }

    /**
     * Serializes the Offering resource to a JSON string.
     *
     * @return The serialized JSON string of the Offering resource.
     * @throws TbdexException if serialization fails.
     */
    fun toJsonString(): String {
        try {
            return rustCoreOffering.toJsonString()
        } catch (e: tbdex.sdk.rust.TbdexException.Exception) {
            throw TbdexException.fromRustCore(e)
        }
    }

    /**
     * Signs the Offering resource using the provided Bearer DID.
     *
     * @param bearerDid The Bearer DID used to sign the Offering resource.
     * @throws TbdexException if the signing process fails.
     */
    fun sign(bearerDid: BearerDid) {
        try {
            rustCoreOffering.sign(RustCoreBearerDid.fromWeb5(bearerDid))
        } catch (e: tbdex.sdk.rust.TbdexException.Exception) {
            throw TbdexException.fromRustCore(e)
        }
    }

    /**
     * Verifies the Offering resource's signature and validity.
     *
     * @throws TbdexException if verification fails.
     */
    fun verify() {
        try {
            rustCoreOffering.verify()
        } catch (e: tbdex.sdk.rust.TbdexException.Exception) {
            throw TbdexException.fromRustCore(e)
        }
    }
}

/**
 * Represents the detailed data for an Offering resource in the tbDEX protocol.
 *
 * @property description A brief description of what is being offered.
 * @property payoutUnitsPerPayinUnit The exchange rate, indicating how many payout units are provided per payin unit.
 * @property payin The details of the payin currency and available methods.
 * @property payout The details of the payout currency and available methods.
 * @property requiredClaims Optional claims required to participate in the offering.
 * @property cancellation The details of the offering's cancellation policy.
 */
data class OfferingData(
    val description: String,
    val payoutUnitsPerPayinUnit: String,
    val payin: PayinDetails,
    val payout: PayoutDetails,
    val requiredClaims: PresentationDefinition?,
    val cancellation: CancellationDetails
)

/**
 * Represents the details of the payin for an Offering.
 *
 * @property currencyCode The ISO 4217 currency code for the payin.
 * @property min Optional minimum payin amount.
 * @property max Optional maximum payin amount.
 * @property methods A list of available methods for making the payin.
 */
data class PayinDetails(
    val currencyCode: String,
    val min: String? = null,
    val max: String? = null,
    val methods: List<PayinMethod>
)

/**
 * Represents a method for making a payin in an Offering.
 *
 * @property kind The kind of payment method.
 * @property name Optional name of the payment method.
 * @property description Optional description of the payment method.
 * @property group Optional group categorization for the payment method.
 * @property requiredPaymentDetails Optional JSON schema specifying required payment details.
 * @property fee Optional fee associated with the payment method.
 * @property min Optional minimum amount for using the payment method.
 * @property max Optional maximum amount for using the payment method.
 */
data class PayinMethod(
    val kind: String,
    val name: String? = null,
    val description: String? = null,
    val group: String? = null,
    val requiredPaymentDetails: Map<String, Any?>? = null,
    val fee: String? = null,
    val min: String? = null,
    val max: String? = null
)

/**
 * Represents the details of the payout for an Offering.
 *
 * @property currencyCode The ISO 4217 currency code for the payout.
 * @property min Optional minimum payout amount.
 * @property max Optional maximum payout amount.
 * @property methods A list of available methods for receiving the payout.
 */
data class PayoutDetails(
    val currencyCode: String,
    val min: String? = null,
    val max: String? = null,
    val methods: List<PayoutMethod>
)

/**
 * Represents a method for receiving a payout in an Offering.
 *
 * @property kind The kind of payout method.
 * @property name Optional name of the payout method.
 * @property description Optional description of the payout method.
 * @property group Optional group categorization for the payout method.
 * @property requiredPaymentDetails Optional JSON schema specifying required payment details.
 * @property fee Optional fee associated with the payout method.
 * @property min Optional minimum amount for using the payout method.
 * @property max Optional maximum amount for using the payout method.
 * @property estimatedSettlementTime The estimated time (in seconds) for the payout to be settled.
 */
data class PayoutMethod(
    val kind: String,
    val name: String? = null,
    val description: String? = null,
    val group: String? = null,
    val requiredPaymentDetails: Map<String, Any?>? = null,
    val fee: String? = null,
    val min: String? = null,
    val max: String? = null,
    val estimatedSettlementTime: Long
)

/**
 * Represents the cancellation policy for an Offering.
 *
 * @property enabled Whether cancellation is enabled for the offering.
 * @property termsUrl Optional URL to a page that describes the terms of cancellation.
 * @property terms Optional human-readable description of the cancellation terms.
 */
data class CancellationDetails(
    val enabled: Boolean,
    val termsUrl: String? = null,
    val terms: String? = null
)